test: reproducers for store mutation with schema change and host down

alecco opened the pull request:
Reproducers for https://github.com/scylladb/scylladb/issues/10770.

(Already fixed in https://github.com/scylladb/scylladb/commit/15ebd5907177ba8709bc40ad2fa9ae773a9cf674)

Includes necessary improvements and fixes to `pylib`.

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/df25f23d889580b7153d84de74d1cefec286f5df:

- Removed change to default number primary keys as it breaks other tests
- commits RF reordered to be together

gleb-cloudius commented on test/topology/test_mutation_schema_change.py:
```diff
@@ -0,0 +1,151 @@
+        change schema + do lwt write + change schema
+        stop B
+        start C
+        do lwt write to the same key through A
```
Through C. The error only in the comment.

gleb-cloudius commented on test/topology/test_mutation_schema_change.py:
```diff
@@ -0,0 +1,151 @@
+# SPDX-License-Identifier: AGPL-3.0-or-later
+#
+"""
+Test repro of failure to store mutation with schema change and a server down
```
The better description of the problem would be:
Reproducer for a failure during lwt operation due to missing of a column mapping in schema history table.

And please include the ref to the fix here 15ebd5907177ba8709bc40ad2fa9ae773a9cf674 which also have a longer description of the problem.

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/307fcc0f761c7321122b7587644274084a76fe3b:

- Fixed comments (Gleb)

gleb-cloudius commented on test/topology/test_mutation_schema_change.py:
```diff
@@ -0,0 +1,150 @@
+    logger.warning("----- STARTING C -----")
+    await manager.server_start(server_c.server_id)
+
+    await asyncio.sleep(10)
```
So did we decide in the end to go with sleeps?

kostja replied:
please no

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4134/

kbr-scylla commented:
If I understand correctly to eliminate the sleeps in this test we need to fix the alive nodes REST API?

How about for now we do what I suggested on Slack earlier:
>  you worked on having a way to reduce the group 0 snapshot threshold through error injection or some other means - I think we should have a simple test that does just this + a bunch of schema changes + adds a node to force a snapshot transfer, and simply expect that nothing breaks during the snapshot transfer. Maybe you could prepare such a test and send it as PR? We don't need the gossiper live nodes thing for that

So I suggest we start with a PR with a **simpler test** that only does the error injection for reducing snapshot threshold, performs a bunch of schema changes, and then bootstraps a node to force a snapshot transfer. This simpler test is not for reproducing https://github.com/scylladb/scylladb/issues/10770, it's only for checking that snapshot transfer works.

Separately we can work on fixing the alive nodes REST API.

Once both things are merged - the reducing snapshot threshold test and the REST API fix, we can go back to this PR and finish it without using sleeps.

gleb-cloudius commented:
> If I understand correctly to eliminate the sleeps in this test we need to fix the alive nodes REST API?
> 
Yes.

> How about for now we do what I suggested on Slack earlier:
> 
> > you worked on having a way to reduce the group 0 snapshot threshold through error injection or some other means - I think we should have a simple test that does just this + a bunch of schema changes + adds a node to force a snapshot transfer, and simply expect that nothing breaks during the snapshot transfer. Maybe you could prepare such a test and send it as PR? We don't need the gossiper live nodes thing for that
> 
> So I suggest we start with a PR with a **simpler test** that only does the error injection for reducing snapshot threshold, performs a bunch of schema changes, and then bootstraps a node to force a snapshot transfer. This simpler test is not for reproducing #10770, it's only for checking that snapshot transfer works.

The test has nothing to do with raft. Why are we discussing snapshot transfer in its context?

kbr-scylla commented:
> The test has nothing to do with raft. Why are we discussing snapshot transfer in its context?

The PR is introducing snapshot threshold error injection and enabling it in the test.

gleb-cloudius commented:
> > The test has nothing to do with raft. Why are we discussing snapshot transfer in its context?
> 
> The PR is introducing snapshot threshold error injection and enabling it in the test.

This is needed for the test to work with raft enabled. But the test may as well ask to disable raft if it is possible instead. The test is not about snapshot transfer.

kbr-scylla commented:
I want to have at least one test which forces snapshot transfer. We have 0 such tests right now. We should not be blocked from introducing a test that forces snapshot transfer by the alive endpoints issue because they are unrelated. Therefore I requested a separate PR with a test that specifically targets snapshot transfer which is not blocked by the alive endpoints issue (as this PR is) but which uses code from this PR (the error injection stuff). I requested it once on Slack and now repeating my request in this PR because this is the first PR that introduces the snapshot threshold error injection.

gleb-cloudius commented:
> I want to have at least one test which forces snapshot transfer. We have 0 such tests right now. We should not be blocked from introducing a test that forces snapshot transfer by the alive endpoints issue because they are unrelated. Therefore I requested a separate PR with a test that specifically targets snapshot transfer which is not blocked by the alive endpoints issue (as this PR is) but which uses code from this PR (the error injection stuff). I requested it once on Slack and now repeating my request in this PR because this is the first PR that introduces the snapshot threshold error injection.

Makes sense.

alecco commented:
> I want to have at least one test which forces snapshot transfer.

https://github.com/scylladb/scylladb/pull/12726

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/7ac9dd46d40cfe8464d334256c682526997d6049:

- rebased on top of https://github.com/scylladb/scylladb/pull/12540 (alive endpoints) (and that one is based on latest master)
- added helper for gossiper endpoints and removed sleeps
- snapshot error injection commits already merged in https://github.com/scylladb/scylladb/pull/12726

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4475/

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/fc06b2acbdd446b5993bf9cccd83f8b495cd2e8c:

- rebased on latest https://github.com/scylladb/scylladb/pull/12540

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4808/

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/856349f5bd81d52424a3ba6648fb27ccc35fe960:

- Parametrize replication factor for topology tests
- Keep using RF 1 for raft upgrade tests

@kbr-scylla

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4842/

kbr-scylla commented on test/topology/conftest.py:
```diff
@@ -216,7 +216,9 @@ def fails_without_consistent_cluster_management(request, check_pre_consistent_cl
-    tables = RandomTables(request.node.name, manager, unique_name())
+    rf_marker = request.node.get_closest_marker("replication_factor")
+    replication_factor = rf_marker.args[0] if rf_marker is not None else 3  # Default 3
+    tables = RandomTables(request.node.name, manager, unique_name(), replication_factor)
```
The default value for `replication_factor` in `RandomTables` is `1`, but there is only one use site of `RandomTables` and it's always passing `replication_factor` and the default is `3`.

Let's remove the default value inside `RandomTables` to make it less confusing.

kbr-scylla commented on test/topology/conftest.py:
```diff
@@ -216,7 +216,9 @@ def fails_without_consistent_cluster_management(request, check_pre_consistent_cl
 # unless the cluster is dirty or the test has failed.
```
Commit message: s/increase/decrease

kbr-scylla commented on test/topology_raft_disabled/test_mutation_schema_change.py:
```diff
@@ -0,0 +1,11 @@
+"""
+Test repro of failure to store mutation with schema change and a server down
+"""
+from test.topology.test_mutation_schema_change import test_mutation_schema_change, \
```
I think running these tests in two suites is unnecessary. The downside is increasing the time to run all tests, in particular CI runs.

Also the tests you made use the `raft_server_snapshot_reduce_threshold` injection, so let's run them only in topology tests only.

kbr-scylla reviewed (changes requested):
How often does each of these two tests reproduce the problem? Like how many times in 100 runs? Let's keep the one which reproduces more often and remove the other one.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4881/

gleb-cloudius commented on test/topology/test_mutation_schema_change.py:
```diff
@@ -0,0 +1,157 @@
+
+
+@pytest.mark.asyncio
+async def test_mutation_schema_change_restart(manager, random_tables):
```
They are similar but trigger two different code paths.

gleb-cloudius commented:
> How often does each of these two tests reproduce the problem? Like how many times in 100 runs? Let's keep the one which reproduces more often and remove the other one.

They trigger the bugs 100% on each run. They trigger two different cases though.

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/f464dab80c851a52d1149f4241b667b5afe0431f:

- rebased on https://github.com/scylladb/scylladb/pull/13240

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/f464dab80c851a52d1149f4241b667b5afe0431f:

- rebased on https://github.com/scylladb/scylladb/pull/13240

alecco commented on gms/gossiper.cc:
```diff
@@ -758,9 +758,12 @@ future<> gossiper::update_live_endpoints_version() {
-    auto live_members = gossiper::get_live_members();
-    co_await replicate_live_endpoints_on_change();
-    co_return live_members;
+    return container().invoke_on(0, [] (gms::gossiper& g) -> future<std::set<inet_address>> {
```
This pr was previously based on [another pr](https://github.com/scylladb/scylladb/pull/12540) that got merged on Friday.

But there was [an issue](https://github.com/scylladb/scylladb/issues/13235) over the weekend. Now there's [another gossiper pr](https://github.com/scylladb/scylladb/pull/13240) with a fix.

And this pr was rebased on top of that pending fix pr. This was mentioned in a comment earlier: https://github.com/scylladb/scylladb/pull/12699#issuecomment-1478254882

nyh replied:
Ok, I see. So we need to wait until this patch goes in and then you rebase this PR.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/205/

alecco commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/205/

Restarted tests.

alecco commented:
> > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/205/
> 
> Restarted tests. (Edit: it was an unrelated failure)

Keeps failing on unrelated 
```
[backup_restore_tests.TestBackupRestore.test_restore_snapshot_using_old_token_ownership](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/205/testReport/junit/backup_restore_tests/TestBackupRestore/Sanity_Tests___test_restore_snapshot_using_old_token_ownership/)
```

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/e322fbfb37df05f5da754cfeff5cbc3c682b305c:

- Reworded change of default replication factor (Kamil)
- Removed default replication factor parameter in `RandomTables` (Kamil)

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/228/

alecco commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/238/

Failed on unrelated debug upgrade test. Rebuilding.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/243/

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/b1484975971676f6ca9544856a43e6aa6fc6ae2f:

- Remove tests with raft disabled (Kamil)

Changes in https://github.com/scylladb/scylladb/commit/040da4e729d06ee8e85d6b37436f4f01792e6561:

- rebase to latest master

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/380/

kbr-scylla commented on test/topology/test_mutation_schema_change.py:
```diff
@@ -0,0 +1,157 @@
+    await wait_for(server_c_sees_another_server, time.time() + 45, period=.1)
+
+    logger.info("Driver connecting to C %s", server_c)
+    manager.ccluster = manager.con_gen([server_c.ip_addr], manager.port, manager.use_ssl)
```
`[server_c.ip_addr]` is only the list of initial contact points for the driver: this is what `cluster_con` (in topology/conftest.py) does:
```python
    return Cluster(execution_profiles={EXEC_PROFILE_DEFAULT: profile},
                   contact_points=hosts,
```
`hosts` is the parameter.

The driver will automatically connect to all alive nodes after querying `system.peers` from the contact points.

So there's no guarantee that the lwt write will go through C as you said in the comment. It could go through A which is also alive now.

You'd have to specify a special `load_balancing_policy` in the `ExecutionProfile` passed to `Cluster` to make sure that the write goes through C.

tgrabiec replied:
I have a patch which allows you to connect to particular coordinators:

https://github.com/scylladb/scylladb/pull/13387/commits/4bea4ec1b71abc3ee9102f104c9d6d94351b42f3

Feel free to commit if good.

kostja replied:
@kbr-scylla @tgrabiec but what is the purpose here of going to a specific node? This is LWT, and if schema is broken on one node, we're going to have failures anyhow.  server_b is down, so LWT has to go to 2 nodes anyway. I don't see any reason to go to server_c specifically with this query.

gleb-cloudius replied:
We test very specific code path which require going to a specific node. We are not just running an LWT query to see it if succeeds.

kostja replied:
I know but it doesn't require that it always goes to the same node. If it goes to a wrong node once in a while the test will succeed. Of course making the test more predictable is valuable, but even without it is good enough. That was my point. 

I tried at some point creating a connection pool that would set up a custom policy, this required very deep refactoring of the current manager code, since connections are opened on demand. 

Anyway @alecco please join the conversation and either address the comment or provide an update to the patch, I don't need to defend the patch for you I think.

gleb-cloudius replied:
> I know but it doesn't require that it always goes to the same node. If it goes to a wrong node once in a while the test will succeed. 

Wont this just produce a fluky test?

alecco replied:
Query goes to the proper node now.

kbr-scylla commented on test/topology/test_mutation_schema_change.py:
```diff
@@ -0,0 +1,157 @@
+    logger.info("Starting C %s", server_c)
+    await manager.server_start(server_c.server_id)
+
+    async def server_c_sees_another_server():
```
Shouldn't we wait for `A` to see another server here, because we'll want to write through `A`?
(not sure I understand why this wait is needed)

kostja replied:
It doesn't make any difference.

alecco replied:
Fixed in latest

kbr-scylla commented on test/topology/test_mutation_schema_change.py:
```diff
@@ -0,0 +1,157 @@
+    await wait_for(server_c_sees_another_server, time.time() + 45, period=.1)
+
+    logger.info("Driver connecting to A %s", server_a)
+    manager.ccluster = manager.con_gen([server_a.ip_addr], manager.port, manager.use_ssl)
```
See previous comment about contact points.

tgrabiec commented on test/topology/test_mutation_schema_change.py:
```diff
@@ -0,0 +1,157 @@
+        alive_nodes = await manager.api.get_alive_endpoints(server_c.ip_addr)
+        if len(alive_nodes) > 1:
+            return True
+    await wait_for(server_c_sees_another_server, time.time() + 45, period=.1)
```
This wait ensures that CQL requests don't fail, right? It's so generally useful, it should be an option to server_start().

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/eb4ffbc968cfe3537895ba83ddf789fae3f1d363:

- Use Tomek's [16075c5ceb](https://github.com/scylladb/scylladb/commit/4bea4ec1b71abc3ee9102f104c9d6d94351b42f3) to force driver to send query to the proper node (Tomek)
- Send query to prope node (Kamil)
- Wait for connections in proper node (Kamil)
- refactor `server_sees_other_server()` (to be moved to pylib later)

Changes in https://github.com/scylladb/scylladb/commit/fc43f10f7e31a6ed1edf1e862ca368c3d99ae0b6:

- rebase to latest master

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/461/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/463/

kbr-scylla commented on test/pylib/random_tables.py:
```diff
@@ -182,8 +182,9 @@ async def add_column(self, name: str = None, ctype: Type[ValueType] = None, colu
             column = Column(name, ctype=ctype)
         self.columns.append(column)
         assert self.manager.cql is not None
-        await self.manager.cql.run_async(f"ALTER TABLE {self.full_name} "
```
The commit message says:

> When adding a new column it's always a value column, not clustering. Fix
the name and sequence.

no such thing is happening in this commit though.

alecco replied:
Yes, another commit was in master. I removed the commit in this series and rebased. Please latest.

kbr-scylla commented on test/topology/conftest.py:
```diff
@@ -216,7 +216,7 @@ def fails_without_consistent_cluster_management(request, check_pre_consistent_cl
 @pytest.fixture(scope="function")
 async def random_tables(request, manager):
-    tables = RandomTables(request.node.name, manager, unique_name())
+    tables = RandomTables(request.node.name, manager, unique_name(), 1)
```
`test_custom` needs to be fixed in a similar way in this `test/pylib: configurable replication factor` commit.

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/4639d5c97a8808f91b040803c051329229bb86b5:

- dropped commit "fix column name" as another version is already in master
- rebase on latest master

kbr-scylla commented on test/pylib/manager_client.py:
```diff
@@ -48,10 +48,11 @@ async def stop(self):
         self.driver_close()
 
-    async def driver_connect(self) -> None:
+    async def driver_connect(self, server=None) -> None:
```
Please include a type hint: `server: Optional[ServerInfo] = None`

alecco replied:
This is Tomek's commit from the tablets series. I wanted to keep it as is so there's no conflict later. 

I'll fix it in [the subsequent PR](https://github.com/scylladb/scylladb/pull/13438).

alecco replied:
done

kbr-scylla commented on test/topology/test_mutation_schema_change.py:
```diff
@@ -0,0 +1,158 @@
+    logger.info("Starting C %s", server_c)
+    await manager.server_start(server_c.server_id)
+
+    await wait_for(partial(server_sees_another_server, server_c, manager), time.time() + 45, period=.1)
```
I don't understand why here you wait only for C to see another server, but in the other test you wait for both servers to see each other.

alecco replied:
The query here goes to C so it needs to know at least of 1 more to have QUORUM.

In the other location it goes through A so it needs to wait for it to know of 1 more for QUORUM. Perhaps the wait for C in the second case is not needed.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/486/

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/1c2ea5d6e090ed38b9daddc07f0b234929ddc03a:

- Re-added wait for C to see A in `test_mutation_schema_change_restart` (due to CI failures)

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/492/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/510/

alecco commented:
Changes in latest:
- Close driver before reconnecting

alecco commented:
Changes in https://github.com/scylladb/scylladb/commit/79ccbfc337dd44d06fa14148ec5030df505cbae8:

- use `wait_for_cql_and_get_hosts()` to avoid driver issue (Kamil)

Changes in https://github.com/scylladb/scylladb/commit/9597822214f891a27b9898f42602c62a677b0958:

- rebase on latest master

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/625/

kbr-scylla commented:
> CI state FAILURE - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/625/

reason for failure wasn't reported
@benipeled could you take a peek?

new build already running: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/632/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/632/

benipeled commented:
> > CI state FAILURE - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/625/
> 
> reason for failure wasn't reported @benipeled could you take a peek?
> 

Failed on dtest https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/637/
The test-result visibility for dtest failures was improved by @tchaikov on https://github.com/scylladb/scylla-pkg/pull/3369 so it should be easier to figure out what's failed (on future runs)
